### PR TITLE
#428: Validate signed URLs regardless of t[] vs t[0]

### PR DESCRIPTION
**Pull request by matslindh**
As detailed in Issue #428, certain sites (Facebook) regenerates URLs on their side where arguments using the array syntax of PHP gets an index appended in the URL. This breaks our signed requests and ends up with Facebook receiving a 400 HTTP error response instead of the image.

By using the Http\QueryString class from Guzzle we can work around the issue by introducing two new URLs to the list of URLs we test when we check the signature. These URLs will have `t[]` replaced with `t[0]` (and up), or `t[0]`, `t[1]`, etc. replaced with `t[]`. This way we'll accept both possible changes.

The choice was made to go with Guzzle's implementation as they have a proper interface to manipulating a query string, which doesn't include regex-ing or manually fixing up the results from `parse_str`, `parse_url` and `http_build_query`. This should be a lot more stable (and we already depend on Guzzle, so there are no new dependencies).

In cases where parameters are intermingled (`t[]=foo&b[]=bar&t[]=baz`) this will not help, but I'm fairly certain we won't see any requests like that in the wild. In addition any request made as `t[1]=foo&t[0]=bar` is undefined (and might give unexpected results).

### CODE REVIEW

File: library/Imbo/EventListener/AccessToken.php [hunk 1]
```diff
@@ -10,9 +10,14 @@
...
 
+use Guzzle\Http\QueryAggregator\QueryAggregatorInterface;
+use Guzzle\Http\Url;
+
 use Imbo\EventManager\EventInterface,
     Imbo\Http\Request\Request,
-    Imbo\Exception\RuntimeException;
+    Imbo\Exception\RuntimeException,
+    Imbo\Helpers\QueryStringArrayConverter,
+    Guzzle\Http\QueryString;
```
Comment (christeredvartsen) [inline library/Imbo/EventListener/AccessToken.php]: Can you change this there is only one `use`?

**Comment by christeredvartsen** [comment]
Today we only depend on Guzzle for development purposes (Behat tests) so we need to move the dependency from `require-dev` to `require` in `composer.json`.

**Comment by christeredvartsen** [comment]
The latest release of Guzzle (`guzzlehttp/guzzle 6.2.0`) uses a different package as well, so please use that one. You will have to update some of your code for it to work with that release.

I'm working on exchanging the version of Guzzle used in the Behat suite with the lastest one as well, so for Imbo-3 we won't have to depend on both versions.

**Comment by matslindh** [comment]
I've rewritten the PR to use `guzzlehttp/psr7` and its helper functions for parsing and building query strings. I've only added `guzzlehttp/psr7` as a dependency for now, which can be removed when the proper dependency on `guzzlehttp/guzzle` is added.

File: library/Imbo/EventListener/AccessToken.php [hunk 2]
```diff
@@ -171,6 +177,88 @@ public function checkAccessToken(EventInterface $event) {
...
+     *
+     * The result for URLs with both formats is undefined, or for URLs that intermingle their parameters,
+     * i.e. t[]=foo&b[]=bar&t[]=baz
+     *
+     * This was introduced because of differences between the URLs generated by the different clients, and
+     * because Facebook (at least) generates URLs were []s in URL arguments are expanded to [0] when
+     * requested from the backend. Since we sign our URLs, this breaks the token generation and thus breaks
+     * URLs when Facebook attempts to retrieve them.
+     *
+     * @param $url string The URL to generate the alternative form of
```
Comment (christeredvartsen) [inline library/Imbo/EventListener/AccessToken.php]: Wrong order, the correct is:

`@param <type> <var> <desc>`

File: library/Imbo/Helpers/Urls.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+namespace Imbo\Helpers;
+
+/**
+ * Helper class for useful functions for building/manipulating URLs
+ *
+ * @author Mats Lindh <mats@lindh.no>
+ * @package Core\Helpers
+ */
+class Urls {
+    public static function buildFromParseUrlParts(array $parts) {
```
Comment (christeredvartsen) [inline library/Imbo/Helpers/Urls.php]: Care to add some phpdoc?

**Comment by matslindh** [comment]
Both phpdoc issues have been fixed.

**Pull request merged**